### PR TITLE
fix: quick fail after pod termination

### PR DESCRIPTION
**Pull request by whynowy**
Fixes: https://github.com/argoproj/argo/issues/1832

When a k8s node where the POD runs has an issue, the POD will go to "terminating" state - which is actually "running" phase but with "DeletionTimestamp", and stuck there. This fix quick fails it when this kind of situation is detected.

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -797,29 +797,36 @@ func assessNodeStatus(pod *apiv1.Pod, node *wfv1.NodeStatus) *wfv1.NodeStatus {
...
-		if err != nil {
-			log.Warnf("%s template annotation unreadable: %v", pod.ObjectMeta.Name, err)
-			return nil
-		}
-		if tmpl.Daemon != nil && *tmpl.Daemon {
-			// pod is running and template is marked daemon. check if everything is ready
-			for _, ctrStatus := range pod.Status.ContainerStatuses {
-				if !ctrStatus.Ready {
-					return nil
+		if pod.DeletionTimestamp != nil {
```
Comment (alexec) [inline workflow/controller/operator.go]: test?

Comment (whynowy) [reply workflow/controller/operator.go]: Added.

**Comment by whynowy** [comment]
Addressing @jessesuen's concern - what happens after setting `activeDeadlineSeconds` in the spec, it might be going through termination process.

`activeDeadlineSeconds` can be set in the spec in 2 places, workflow level or template level.
See following example: 
```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: sleep-
spec:
  ##### activeDeadlineSeconds: 30
  entrypoint: sleep-n-sec
  arguments:
    parameters:
    - name: seconds
      value: 60
  templates:
  - name: sleep-n-sec
    #####   activeDeadlineSeconds: 30
    inputs:
      parameters:
      - name: seconds
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["echo sleeping for {{inputs.parameters.seconds}} seconds; sleep {{inputs.parameters.seconds}}; echo done"]
```

a.) If it's set in workflow level, when it comes to the deadline, the `main` container will be killed by the `wait` container through `docker kill --signal TERM xxxxx`, and then the pod goes to `Error` status (`failed` phase), it's not through termination process, thus `DeletionTimestamp` is still `nil`, the workflow goes to `failed`.

b.) If `activeDeadlineSeconds` is set on the template level, it will be translated to POD spec, when the time comes, the pod will be set to `DeadlineExceeded` status (still `failed` phase), not going through termination process as well, the workflow also goes to `failed`.

In summary, there's no impact for existing features with this change.

**Comment by alexec** [comment]
@jessesuen do you want to review this?

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -797,29 +797,36 @@ func assessNodeStatus(pod *apiv1.Pod, node *wfv1.NodeStatus) *wfv1.NodeStatus {
...
+			// pod is being terminated
+			newPhase = wfv1.NodeError
+			message = "pod termination"
+			log.Info(message)
```
Comment (alexec) [inline workflow/controller/operator.go]: Logging should really be structured, so we can create metrics and dashboard, could you make sure this uses `log.WithFields(log.Fields{}).Info("")` to log the workflow name please?

**Comment by alexec** [comment]
I've taken the liberty of syncing you with master so you have the new test infra.

Would you like to add an e2e test for this?

File: workflow/controller/operator.go (hunk 2 above)
Comment (whynowy) [reply workflow/controller/operator.go]: Don't need to, it has been set as default in https://github.com/argoproj/argo/blob/master/workflow/controller/operator.go#L110-L112

Comment (whynowy) [reply workflow/controller/operator.go]: My bad, ignoring previous comment. I'll make change.

Comment (whynowy) [reply workflow/controller/operator.go]: I simply removed the redundant log, since there are 2 other logs for "node update" and "message update" around, and the "message" is already included in them. 

The best way to use `log` in this function is to turn it to a struct function so that the `log` in the struct can be used, which already has `WithFields()`.

**Comment by whynowy** [comment]
@alexec - e2e test added, could you please review it again?

File: test/e2e/functional_test.go [hunk 3]
```diff
@@ -69,6 +69,22 @@ spec:
 		})
 }
 
+func (s *FunctionalSuite) TestFastFailOnPodTermination() {
+	s.Given().
+		Workflow("@expectedfailures/pod-termination-failure.yaml").
+		When().
+		SubmitWorkflow().
+		WaitForWorkflow(120 * time.Second).
```
Comment (alexec) [inline test/e2e/functional_test.go]: minor - change to 60s

**Comment by salanki** [comment]
Thanks to everyone involved!
